File: 심수연/4주차/260121.py
# https://www.acmicpc.net/problem/2156

import sys
input = sys.stdin.readline

n = int(input())
wine = [int(input()) for _ in range(n)]

dp = [0] * n

# dp[i] = i번째 잔까지 고려했을 때 마실 수 있는 최대 포도주 양

dp[0] = wine[0] # 첫 잔은 마시는 게 최댓값

if n > 1:
  dp[1] = wine[0] + wine[1] # 두 잔 가능

if n > 2:
  dp[2] = max(
    wine[0] + wine[1], # 2 빼고 먹기
    wine[0] + wine[2], # 1 빼고 먹기
    wine[1] + wine[2], # 0 빼고 먹기
    ) 

for i in range(3, n):
  dp[i] = max(
    dp[i - 1], # i번째 잔 안 마시기
    dp[i - 2] + wine[i], # i-2 (O), i-1 (X), i (O)
    dp[i - 3] + wine[i - 1] + wine[i] # i-3 (O), i-2 (X), i-1 (O), i (O)
    )
  
print(dp[n - 1]) # 배열이 0부터 시작하니까 n - 1

File: 박예진/4주차/260121.cpp
//https://www.acmicpc.net/problem/2234

#include <iostream>
#include <algorithm>
#include <vector>
#include <queue>
using namespace std;

struct Node {
  int x, y;
};

int N, M, arr[51][51];
bool visited[51][51];
int roomIdx[51][51];
// 서, 북, 동, 남
const int dx[4] = {0, -1, 0, 1};
const int dy[4] = {-1, 0, 1, 0};

bool OOB(int x, int y){
  if (x < 0 || x >= N || y < 0 || y >= M) return true;
  return false;
}

int bfs(int x, int y, int idx) {
  queue<Node> q;
  q.push({x, y});
  visited[x][y] = true;
  int cnt = 0;

  while(!q.empty()) {
    Node now = q.front(); q.pop();
    roomIdx[now.x][now.y] = idx;
    cnt++;

    for(int dir = 0; dir < 4; dir++){
      int nx = now.x + dx[dir];
      int ny = now.y + dy[dir];

      if (OOB(nx, ny) || visited[nx][ny]) continue;
      if (arr[now.x][now.y] & (1 << dir)) continue; // 벽이 있다면
      q.push({nx, ny});
      visited[nx][ny] = true;
    }
  }
  return cnt;
}

int main(){
  cin >> M >> N;
  for(int i = 0; i < N; i++){
    for(int j = 0; j < M; j++){
      cin >> arr[i][j];
    }
  }

  // bfs 돌려서 방의 개수, 넓이 구하기
  int roomCnt = 0;
  int roomSize = 0;
  int res = 0;
  vector<int> rooms;

  for(int i = 0; i < N; i++){
    for(int j = 0; j < M; j++){
      if (!visited[i][j]) {
        int num = bfs(i, j, roomCnt);
        roomCnt++;
        roomSize = max(roomSize, num);
        rooms.push_back(num);
      }
    }
  }

  // 각 연결되어 있는 방 검색
  for(int i = 0; i < N; i++){
    for(int j = 0; j < M; j++){
      for(int dir = 0; dir < 4; dir++){
        int nx = i + dx[dir];
        int ny = j + dy[dir];

        if (OOB(nx, ny)) continue;
        if (roomIdx[i][j] != roomIdx[nx][ny]) {
          res = max(res, rooms[roomIdx[i][j]] + rooms[roomIdx[nx][ny]]);
        }
      }
    }
  }

  cout << roomCnt << "\n" << roomSize << "\n" << res;
}